models/user.go

```diff
@@ -1652,14 +1659,50 @@ func (opts *SearchUserOptions) toConds() builder.Cond {
 		cond = cond.And(builder.Eq{"is_active": opts.IsActive.IsTrue()})
 	}
 
-	return cond
+	if !opts.IsAdmin.IsNone() {
+		cond = cond.And(builder.Eq{"is_admin": opts.IsAdmin.IsTrue()})
+	}
+
+	if !opts.IsRestricted.IsNone() {
+		cond = cond.And(builder.Eq{"is_restricted": opts.IsRestricted.IsTrue()})
+	}
+
+	if !opts.IsProhibitLogin.IsNone() {
+		cond = cond.And(builder.Eq{"prohibit_login": opts.IsProhibitLogin.IsTrue()})
+	}
+
+	type Join struct {
+		Operator  string
+		Table     interface{}
+		Condition string
+		Args      []interface{}
+	}
+
+	var joins []*Join
+
+	if !opts.IsTwoFactorEnabled.IsNone() {
+		// 2fa filter uses LEFT JOIN to check whether a user has a 2fa record
+		// TODO: bad performance here, maybe there will be a column "is_2fa_enabled" in the future
+		if opts.IsTwoFactorEnabled.IsTrue() {
+			cond = cond.And(builder.Expr("two_factor.uid IS NOT NULL"))
+		} else {
+			cond = cond.And(builder.Expr("two_factor.uid IS NULL"))
+		}
+		joins = append(joins, &Join{Operator: "LEFT OUTER", Table: "two_factor", Condition: "two_factor.uid = `user`.id"})
+	}
```

Comment:
You should be able to use ```go
sess = x.NewSession()
```
at the start of the function then just use `sess = sess.Join(...)`  in here and `sess.Where` below.

Reply:
It's very difficult to write such code in current framework (maybe I haven't fully understood the xorm framework)

My question is: how to get the first session object clearly?

Sometimes we need to join then start with `x.Join`, sometimes we just need to start with `x.Where`.

And I do not want to make `x.Where` or `x.Join` appear between `cond` logic, it's too hard to understand.

Reply:
So what I did here, is to prepare all variables for `Where` and `Join` first, then make a full query use prepared data.

Even if there will be for `join`s in future, the logic stays the same.

Reply:
what you are doing is essentially creating your own version of a session and placing the join between the cond logic.

You should either use the session or retest the optional bool later on.

Reply:
> My question is: how to get the first session object clearly?

```go
sess = x.NewSession()
```

Reply:
ps: why I didn't use `NewSession`: it doesn't have `session.isAutoClose = true` as `Where` or `Join`, I am not sure whether it works correctly.  If you can confirm that `NewSession` works, I can rewrite.

Reply:
ps again:

such code doesn't work in xorm:  `Count: Error 1054: Unknown column 'two_factor.uid' in 'where clause'`

```

	sess = x.NewSession()
	if !opts.IsTwoFactorEnabled.IsNone() {
		// 2fa filter uses LEFT JOIN to check whether a user has a 2fa record
		// TODO: bad performance here, maybe there will be a column "is_2fa_enabled" in the future
		if opts.IsTwoFactorEnabled.IsTrue() {
			cond = cond.And(builder.Expr("two_factor.uid IS NOT NULL"))
		} else {
			cond = cond.And(builder.Expr("two_factor.uid IS NULL"))
		}
		sess = sess.Join("LEFT OUTER", "two_factor", "two_factor.uid = `user`.id")
	}
	sess = x.Where(cond)
```

Reply:
Sorry, I know what's wrong here.